go/vt/vtgateproxy/discovery.go

```diff
@@ -48,225 +41,312 @@ var (
+			targets = append(targets, target)
+			log.V(1000).Infof("matched target %v with type %s", target, r.poolType)
+		} else {
+			log.V(1000).Infof("skipping host %v with type %s", target, r.poolType)
 		}
 	}
 
-	// Grab the first N addresses, and voila!
-	var addrs []resolver.Address
-	hosts = hosts[:min(*numConnections, len(hosts))]
-	for _, host := range hosts {
-		addrs = append(addrs, resolver.Address{
-			Addr: fmt.Sprintf("%s:%s", host[*addressField], host[*portField]),
-		})
+	// Shuffle to ensure every host has a different order to iterate through, putting
+	// the affinity matching (e.g. same az) hosts at the front and the non-matching ones
+	// at the end.
+	//
+	// Only need to do n-1 swaps since the last host is always in the right place.
+	n := len(targets)
+	head := 0
+	tail := n - 1
+	for i := 0; i < n-1; i++ {
+		j := head + b.rand.Intn(tail-head+1)
+
+		if r.affinity == "" || r.affinity == targets[j].affinity {
+			targets[head], targets[j] = targets[j], targets[head]
+			head++
+		} else {
+			targets[tail], targets[j] = targets[j], targets[tail]
+			tail--
+		}
 	}
 
-	h := sha256.New()
-	if _, err := io.Copy(h, bytes.NewReader(data)); err != nil {
-		return nil, nil, err
+	// Grab the first N addresses, and voila!
+	var addrs []resolver.Address
+	targets = targets[:min(*numConnections, len(targets))]
+	for _, target := range targets {
+		addrs = append(addrs, resolver.Address{Addr: target.addr})
 	}
-	sum := h.Sum(nil)
 
-	log.V(100).Infof("resolved %s to hosts %s addrs: 0x%x, %v\n", r.target.URL.String(), jsonDump(hosts), sum, addrs)
+	log.V(100).Infof("updated targets for %s to %v", r.target.URL.String(), targets)
 
-	return &addrs, sum, nil
+	r.clientConn.UpdateState(resolver.State{Addresses: addrs})
 }
 
-func (r *JSONGateConfigResolver) start() {
-	log.V(100).Infof("Starting discovery checker\n")
-	r.rand = rand.New(rand.NewSource(time.Now().UnixNano()))
+// Build a new Resolver to route to the given target
+func (b *JSONGateResolverBuilder) Build(target resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions) (resolver.Resolver, error) {
+	attrs := target.URL.Query()
 
-	// Immediately load the initial config
-	addrs, hash, err := r.resolve()
-	if err == nil {
-		// if we parse ok, populate the local address store
-		r.cc.UpdateState(resolver.State{Addresses: *addrs})
+	// If the config specifies a pool type attribute, then the caller must supply it in the connection
+	// attributes, otherwise reject the request.
+	poolType := ""
+	if b.poolTypeField != "" {
+		poolType = attrs.Get(b.poolTypeField)
+		if poolType == "" {
+			return nil, fmt.Errorf("pool type attribute %s not in target", b.poolTypeField)
+		}
 	}
 
-	// Start a config watcher
-	r.ticker = time.NewTicker(100 * time.Millisecond)
-	fileStat, err := os.Stat(r.jsonPath)
-	if err != nil {
-		return
+	// Affinity on the other hand is just an optimization
+	affinity := ""
+	if b.affinityField != "" {
+		affinity = attrs.Get(b.affinityField)
```

Comment:
Do I have this right: grpc will call `Build` when a new client connection is created, which is identified by the `vtgate://<type>` portion of the URL. 

So if we ask for the same `<type>` but with two different affinity attributes, we will only create one resolver corresponding to the first affinity.

This likely works in practice because why would we change the affinity from the local AZ? But it seems more correct to defer affinity to the actual subconn picking step (which would probably mean implementing a custom load balancer).

Reply:
>But it seems more correct to defer affinity to the actual subconn picking step (which would probably mean implementing a custom load balancer).

That's how @jscheinblum originally implemented it before the refactor that preceded this refactor.

The problem is that GRPC will connect to every host that you return from the discovery list. So you end up with all these actually unused TCP connections. So, we had to move the affinity check and the subsetting into the discovery phase and not the load balancing phase.

Reply:
> Do I have this right: grpc will call Build when a new client connection is created, which is identified by the vtgate://<type> portion of the URL.

Not quite -- Build will be called based on the full URL uniqueness. So even if you change things so there's a different az_id or we omit az_id altogether, it will get a new set of hosts for that target.

I just verified this in dev and tweaked the metrics a bit so that if we ever don't know the az_id of the target host we now track it as "unknown"